src/hostname_resolution.py

```diff
@@ -0,0 +1,125 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Library containing logic pertaining to hostname resolutions in the VM charm."""
+
+import io
+import logging
+import socket
+
+from ops.charm import CharmBase
+from ops.framework import Object
+from ops.model import Unit
+
+from constants import HOSTNAME_TO_IP_KEY, PEER
+
+logger = logging.getLogger(__name__)
+
+
+class MySQLMachineHostnameResolution(Object):
+    """Encapsulation of the the machine hostname resolution."""
+
+    def __init__(self, charm: CharmBase):
+        super().__init__(charm, "hostname-resolution")
+
+        self.charm = charm
+
+        self.framework.observe(self.charm.on.config_changed, self._update_hostname_ip_in_databag)
+
+        self.framework.observe(
+            self.charm.on[PEER].relation_changed, self._potentially_update_etc_hosts
+        )
+
+    def _update_hostname_ip_in_databag(self, _) -> None:
+        with open("/etc/hostname", "r") as hostname_file:
+            hostname = hostname_file.readline().strip()
+
+        def _get_local_ip():
+            s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+            s.settimeout(0)
+
+            try:
+                s.connect(("10.10.10.10", 1))
+                ip = s.getsockname()[0]
+            except Exception:
+                logger.exception("Unable to get local IP address")
+                ip = "127.0.0.1"
```

Comment:
Just a question: why don't we use IP provided by Juju in databad and trying to guess "my IP"? 
> private-address: 10.18.215.88

Reply:
When `config-changed` or `peer-relation-joined` events fire, the local unit databag is not always populated (the IP address and hostname of the unit gets added to the databag at some point, but does not trigger a `peer-relation-changed`). Instead of dealing with defers and other mechanisms, I thought that using `socket` to get the IP address of the instance would be cleaner